src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->decreaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentAdded(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->increaseTorrentsCount();
+
+    const QVector<TagModelItem *> items = findItems(torrent->tags());
+    if (items.isEmpty())
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    foreach (TagModelItem *item, items)
+        item->increaseTorrentsCount();
+}
+
+void TagFilterModel::torrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->decreaseTorrentsCount();
+
+    if (torrent->tags().isEmpty())
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    foreach (TagModelItem *item, findItems(torrent->tags()))
+        item->decreaseTorrentsCount();
+}
+
+QString TagFilterModel::getSpecialAllTag()
+{
+    static const QString *const ALL_TAG = new QString(" ");
+    Q_ASSERT(!BitTorrent::Session::isValidTag(*ALL_TAG));
+    return *ALL_TAG;
+}
+
+QString TagFilterModel::getSpecialUntaggedTag()
+{
+    static const QString *const UNTAGGED_TAG = new QString("  ");
+    Q_ASSERT(!BitTorrent::Session::isValidTag(*UNTAGGED_TAG));
+    return *UNTAGGED_TAG;
+}
+
+void TagFilterModel::populate()
+{
+    using Torrent = BitTorrent::TorrentHandle;
+
+    auto session = BitTorrent::Session::instance();
+    auto torrents = session->torrents();
+
+    // All torrents
+    addItemToModel(new TagModelItem(getSpecialAllTag(), tr("All"), torrents.count()));
+
+    const int untaggedCount = std::count_if(torrents.begin(), torrents.end(),
+                                             [](Torrent *torrent) { return torrent->tags().isEmpty(); });
+    addItemToModel(new TagModelItem(getSpecialUntaggedTag(), tr("Untagged"), untaggedCount));
+
+    foreach (const QString &tag, session->tags()) {
+        const int count = std::count_if(torrents.begin(), torrents.end(),
+                                        [tag](Torrent *torrent) { return torrent->hasTag(tag); });
+        addItemToModel(new TagModelItem(tag, tag, count));
+    }
+}
+
+void TagFilterModel::addItemToModel(TagModelItem *item)
+{
+    Q_ASSERT(item);
+    const QString &tag = item->tag();
+    Q_ASSERT(!m_tagRows.contains(tag));
+    m_tagRows.append(tag);
+    m_tagItems[tag] = item;
+}
+
+void TagFilterModel::removeItemFromModel(TagModelItem *item)
+{
+    Q_ASSERT(item);
+    const QString &tag = item->tag();
+    m_tagRows.removeOne(tag);
+    m_tagItems.remove(tag);
+    delete item;
+}
+
+TagModelItem *TagFilterModel::findItem(const QString &tag) const
+{
+    return m_tagItems.value(tag, nullptr);
+}
+
+QVector<TagModelItem *> TagFilterModel::findItems(const QSet<QString> &tags) const
+{
+    QVector<TagModelItem *> items;
+    foreach (const QString &tag, tags) {
+        TagModelItem *item = findItem(tag);
+        if (item)
```

Comment:
Yes, that would indicate some sort of internal inconsistency. I added a QWarning.